src/toxcore/onion/onion_announce.rs

```diff
@@ -0,0 +1,521 @@
+`2`      | Port of sender
+
+*/
+struct OnionPingData {
+    /// Secret bytes of onion node to make onion ping id unique
+    pub secret_bytes: [u8; SECRET_BYTES_SIZE],
+    /// Unix time in seconds
+    pub time: u64,
+    /// `PublicKey` of sender
+    pub pk: PublicKey,
+    /// `IpAddr` of sender
+    pub ip_addr: IpAddr,
+    /// Port of sender
+    pub port: u16
+}
+
+impl ToBytes for OnionPingData {
+    fn to_bytes<'a>(&self, buf: (&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError> {
+        do_gen!(buf,
+            gen_slice!(&self.secret_bytes) >>
+            gen_be_u64!(self.time / PING_ID_TIMEOUT) >>
+            gen_slice!(self.pk.as_ref()) >>
+            gen_be_u8!(self.ip_addr.is_ipv4() as u8) >>
+            gen_call!(|buf, ip_addr| IpAddr::to_bytes(ip_addr, buf), &self.ip_addr) >>
+            gen_be_u16!(self.port)
+        )
+    }
+}
+
+impl OnionPingData {
+    pub fn ping_id(&self) -> Digest {
+        let mut buf = [0; ONION_PING_DATA_SIZE];
+        // can not fail since buf has enough length
+        self.to_bytes((&mut buf, 0)).unwrap();
+        hash(&buf)
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct OnionAnnounce {
+    secret_bytes: [u8; SECRET_BYTES_SIZE],
+    entries: Vec<OnionAnnounceEntry>,
+    dht_pk: PublicKey
+}
+
+impl OnionAnnounce {
+    pub fn new(dht_pk: PublicKey) -> OnionAnnounce {
+        let mut secret_bytes = [0; SECRET_BYTES_SIZE];
+        randombytes_into(&mut secret_bytes);
+        OnionAnnounce {
+            secret_bytes,
+            entries: Vec::with_capacity(ONION_ANNOUNCE_MAX_ENTRIES),
+            dht_pk
+        }
+    }
+
+    pub fn ping_id(&self, time: u64, pk: PublicKey, ip_addr: IpAddr, port: u16) -> Digest {
+        let data = OnionPingData {
+            secret_bytes: self.secret_bytes,
+            time,
+            pk,
+            ip_addr,
+            port
+        };
+        data.ping_id()
+    }
+
+    fn in_entries(&self, pk: PublicKey) -> Option<&OnionAnnounceEntry> {
+        self.entries.iter().find(|e| !e.is_timed_out() && e.pk == pk)
+    }
+
+    /** Try to add announce entry to onion announce list.
+
+    Firstly we remove all timed out entries. Then if:
+    - announce list already contains entry with such `PublicKey` then update
+      entry and return it
+    - announce list with new entry does not exceed `ONION_ANNOUNCE_MAX_ENTRIES`
+      length add entry to the list and return it
+    - the farthest entry from DHT `PublicKey` is farther than new entry then
+      replace it with new entry
+
+    Also we keep onion announce list sorted by distance to DHT `PublicKey` so
+    we can easily find the farthest entry.
+
+    */
+    fn add_to_entries(&mut self, entry: OnionAnnounceEntry) -> Option<&OnionAnnounceEntry> {
+        let dht_pk = &self.dht_pk;
+        self.entries.retain(|e| !e.is_timed_out());
+        //TODO: replace with `self.entries.iter_mut().find(|e| e.pk == entry.pk` when NLL get stabilized
```

Comment:
You may clone entry.pk =)

Reply:
I can't because I need to update/return reference to entry therefore it will borrow the whole entries vector outside of if-block.